### ISSUE TITLE
Support clip path hit testing drop target without clipping children.

### ISSUE THREAD

**Issue by flackr** [ArnoldSmith86/virtualtabletop#1572]
@robartsd This may be covered by #984 but I thought it would be a good test case to make sure and give an example of the use case.

Here's a sample (based on a real game) where pieces are attached to irregular rotated board pieces:
https://virtualtabletop.io/clip-path-holder

The clip-path ensures that the piece is dropped on the correct holder and that the correct holder is dragged however it also clips the children. It would also be nice if the pieces weren't rotated when dropped but that's what #1388 is supposed to help with.

Note there also seems to be a bug when a drag is started that the position is incorrect :(.

**Comment by robartsd** [comment]
I'm pretty sure that #984 will close this issue as designed already, but have no problem making closing this issue a requirement for that PR.

**Comment by flackr** [comment]
Sounds good. The other case (which I think will already be supported) is imagine my example but perhaps you don't want to actually clip the board graphic because it has some sort of a glow or something else. I imagine this would work by putting the image as a background of the top object instead of the decoration layer?

**Comment by flackr** [comment]
In particular, my proposal would be that we have a property hitTestPath or something like that, which would have the exact format of `clip-path` and when set would result in a hit test element being created and imply `pointer-events: none` on the presentation element. In most cases game authors shouldn't need to directly style the hit testing element.

**Comment by robartsd** [comment]
That would work on the widget being dragged; however, this issue is about the holder. Widget.move() uses Document.elementsFromPoint() to find potential holders under the widget. Document.elementsFromPoint() does not care what the computed pointer-events style of the elements are. If all widgets don't have a hit test element, we will have to deal with both appropriately in Widget.move().

**Comment by flackr** [comment]
> That would work on the widget being dragged; however, this issue is about the holder. Widget.move() uses Document.elementsFromPoint() to find potential holders under the widget. Document.elementsFromPoint() does not care what the computed pointer-events style of the elements are. If all widgets don't have a hit test element, we will have to deal with both appropriately in Widget.move() (eliminating widgets at the point that have a hit test element that is not at the point and adding widgets that have a hit test element at the point even if the widget itself is not at the point).

I think you're mistaken, elementsFromPoint does respect `pointer-events: none` in Chrome, Safari and Firefox. Demo at 
 https://jsbin.com/mogumar/edit?html,css,js,output

**Comment by robartsd** [comment]
The documentation at [https://developer.mozilla.org/en-US/docs/Web/API/Document/elementsFromPoint](https://developer.mozilla.org/en-US/docs/Web/API/Document/elementsFromPoint) does not make that clear (it is clear that elementFromPoint() does test pointer-events). I thought elementsFromPoint() would return all elements and elementFromPoint() would return the first element returned by elementsFromPoint() that accepts pointer events.

This means the elementsFromPoint() won't work to find hidden elements in JSON edit mode.

**Comment by flackr** [comment]
That's why in that PR I temporarily add the `.hitTest` class:
https://github.com/ArnoldSmith86/virtualtabletop/pull/1571/files#diff-5f89c6936ec8924275ac03ec4dfcf4b77c3c497c105f3fe8ef50fffaa1e15733R2429

Which removes `pointer-events: none`:
https://github.com/ArnoldSmith86/virtualtabletop/pull/1571/files#diff-96478b88071667c8d36a06b2f2eff5bb2d771bdeaafff9b241f3c582c4db123dR672

**Comment by robartsd** [comment]
Even the [draft specification ](https://w3c.github.io/csswg-drafts/cssom-view/#dom-document-elementsfrompoint)is not very clear:

> The elementsFromPoint(x, y) method must follow these steps:
> 1. Let sequence be a new empty sequence.
> 2. If either argument is negative, x is greater than the [viewport](https://w3c.github.io/csswg-drafts/css2/#viewport%E2%91%A0) width excluding the size of a rendered scroll bar (if any), or y is greater than the viewport height excluding the size of a rendered scroll bar (if any), or there is no viewport associated with the document, return sequence and terminate these steps.
> 3. For each [box](https://w3c.github.io/csswg-drafts/css-display-4/#box) in the [viewport](https://w3c.github.io/csswg-drafts/css2/#viewport%E2%91%A0), in paint order, starting with the topmost box, that would be a target for hit testing at coordinates x,y even if nothing would be overlapping it, when applying the [transforms](https://w3c.github.io/csswg-drafts/cssom-view/#transforms) that apply to the descendants of the viewport, append the associated element to sequence.
> 4. If the document has a root element, and the last item in sequence is not the root element, append the root element to sequence.
> 5. Return sequence.

Step 3 (actually telling which elements to select) is difficult to parse. It corresponds to step 2 in the steps for elementFromPoint()

> 2. If there is a [box](https://w3c.github.io/csswg-drafts/css-display-4/#box) in the [viewport](https://w3c.github.io/csswg-drafts/css2/#viewport%E2%91%A0) that would be a target for hit testing at coordinates x,y, when applying the [transforms](https://w3c.github.io/csswg-drafts/cssom-view/#transforms) that apply to the descendants of the viewport, return the associated element and terminate these steps.

The note for elementFromPoint() does make it clear that pointer-events: none excludes elements from being hit testing targets (which applies to step 3 for elementsFromPoint()).

> NOTE: The [elementFromPoint()](https://w3c.github.io/csswg-drafts/cssom-view/#dom-document-elementfrompoint) method does not necessarily return the top-most painted element. For instance, an element can be excluded from being a target for hit testing by using the [pointer-events](https://w3c.github.io/csswg-drafts/css-ui-4/#propdef-pointer-events) CSS property.

**Comment by flackr** [comment]
Yeah exactly. It's not very obvious from the spec, the key is it says "that would be a target for hit testing at coordinates x,y". `pointer-events: none` is not a target for hit testing.

**Comment by robartsd** [comment]
> Yeah exactly. It's not very obvious from the spec, the key is it says "that would be a target for hit testing at coordinates x,y". `pointer-events: none` is not a target for hit testing.

Part of my problem with the language is "if nothing would be overlapping it". I think I'd rather it say "ignoring any overlapping elements".

It would be nice if there were a way to get elements at a point that included elements that are not targets for hit testing; but I suppose the spec wants user agents to be able to use whatever hit testing optimizations they already have when finding elementsFromPoint.

**Comment by flackr** [comment]
> > Yeah exactly. It's not very obvious from the spec, the key is it says "that would be a target for hit testing at coordinates x,y". `pointer-events: none` is not a target for hit testing.
> 
> Part of my problem with the language is "if nothing would be overlapping it". I think I'd rather it say "ignoring any overlapping elements".

I agree, I suspect the current language may not have been carefully read over. Do you want to send a PR?

> It would be nice if there were a way to get elements at a point that included elements that are not targets for hit testing;

Agreed. I searched and couldn't find any pre-existing issues for this. Want to file one? I could see adding an options dictionary to elementsFromPoint that allowed getting elements regardless of if they are candidate hit targets.

> but I suppose the spec wants user agents to be able to use whatever hit testing optimizations they already have when finding elementsFromPoint.

I highly doubt the spec was written this way for performance, the [priority of constituencies](https://www.w3.org/TR/design-principles/#priority-of-constituencies) when making tradeoffs is to first prioritize the users, then web developers, and lastly browsers. Most likely it was to try to avoid author surprise if elementsFromPoint returned elements that weren't actually event targets.

**Comment by robartsd** [comment]
I don't know much about the w3 specification writing process. I suppose contributing to that process would be a good idea.

**Comment by RaphaelAlvez** [comment]
Why is there no test server?

**Comment by flackr** [comment]
> Why is there no test server?

This is an issue, not a PR.

**Comment by RaphaelAlvez** [comment]
> > Why is there no test server?
> 
> This is an issue, not a PR.

🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️

**Comment by ArnoldSmith86** [comment]
Still working on the AI that generates test servers for the issues.
